test/e2e/networkpolicy_test.go

```diff
@@ -0,0 +1,95 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"testing"
+)
+
+func TestIPBlockWithExcept(t *testing.T) {
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+
+	workerNode := workerNodeName(1)
+
+	podName0 := randName("test-pod-networkpolicy-")
+	if err := data.createBusyboxPodOnNode(podName0, workerNode); err != nil {
+		t.Fatalf("Error when creating busybox test Pod: %v", err)
+	}
+	defer deletePodWrapper(t, data, podName0)
+	if _, err := data.podWaitForIP(defaultTimeout, podName0); err != nil {
+		t.Fatalf("Error when waiting for IP for Pod '%s': %v", podName0, err)
+	}
+
+	podName1 := randName("test-pod-networkpolicy-")
+	if err := data.createBusyboxPodOnNode(podName1, workerNode); err != nil {
+		t.Fatalf("Error when creating busybox test Pod: %v", err)
+	}
+	defer deletePodWrapper(t, data, podName1)
+	podIP1, err := data.podWaitForIP(defaultTimeout, podName1)
+	if err != nil {
+		t.Fatalf("Error when waiting for IP for Pod '%s': %v", podName1, err)
+	}
+
+	nginxPodName := randName("test-pod-nginx-")
+	if err = data.createNginxPodOnNode(nginxPodName, workerNode); err != nil {
+		t.Fatalf("Error when creating nginx pod: %v", err)
+	}
+	defer deletePodWrapper(t, data, nginxPodName)
+	if _, err := data.podWaitForIP(defaultTimeout, nginxPodName); err != nil {
+		t.Fatalf("Error when waiting for IP for Pod '%s': %v", nginxPodName, err)
+	}
+
+	svcName := randName("test-svc-nginx")
+	if err = data.createNginxService(svcName); err != nil {
+		t.Fatalf("Error when creating nginx service: %v", err)
+	}
+	defer func() {
+		if err = data.deleteService(svcName); err != nil {
+			t.Fatalf("delete nginx service error: %v", err)
+		}
+	}()
+
+	// Two pods can wget to service.
+	if err = data.runWgetCommandFromTestPod(podName0, svcName); err != nil {
+		t.Fatalf("Error when %s runs wget, it should not fail: %v", podName0, err)
+	}
+	if err = data.runWgetCommandFromTestPod(podName1, svcName); err != nil {
+		t.Fatalf("Error when %s runs wget, it should not fail: %v", podName1, err)
+	}
+
+	networkPolicyName := "networkpolicy-test-ipblock-except"
+	createdPolicy, err := data.createNetworkPolicyToTestIPBlockWithExcept(networkPolicyName, podIP1)
+	if err != nil {
+		t.Fatalf("Error when creating network policy: %v", err)
+	}
+	defer func() {
+		if err = data.deleteNetworkpolicy(createdPolicy); err != nil {
+			t.Fatalf("Error when deleting network policy: %v", err)
+		}
+	}()
+
+	// Now pod0 can still wget to service.
+	if err = data.runWgetCommandFromTestPod(podName0, svcName); err != nil {
+		t.Fatalf("Pod %s should be able to connect Service %s, but was not able to connect", podName0, svcName)
+	}
+	// Now pod1 cannot wget to service.
+	if err = data.runWgetCommandFromTestPod(podName1, svcName); err == nil {
```

Comment:
is it not possible to get a race condition here? When the call to `createNetworkPolicyToTestIPBlockWithExcept` returns, there is no guarantee that the flows have been installed already (@tnqn is that not the case?). Then we could get in a situation where the `wget` succeeds here.

How about:
1) first checking that pod1 cannot wget using `wait.PollImmediate` and a small timeout
2) then checking that pod0 can still wget

Reply:
Quan suggested that there is enough time for the networkpolicy to be deployed. Do we need a timeout for guarantee @tnqn ?

I agree to put pod1 before pod0 here.

Reply:
I think a race condition here is quite likely, especially in the light of https://github.com/vmware-tanzu/antrea/issues/197. I would suggest adding the "busy wait" here, at least until #197 is fixed to avoid spurious failures in the test.

Reply:
The race condition only exists in the situation that the Pod tests its egress connection before its PodIP is updated in K8s APIServer, which is because the Pod's command (Pid 1) is the connection test command. For ingress cases, I haven't seen the condition happened, which should mean waiting the Create request return and remotely executing a command in a Pod is long enough for antrea components to enforce the policy. 
And the K8s NetworkPolicy ingress tests (e.g. https://github.com/kubernetes/kubernetes/blob/ed8a260fb3ff03843e3e29bc8e0f24fe4bfb7ab4/test/e2e/network/network_policy.go#L765) don't wait for the enforcement and we have passed all, I think we could follow the same criteria.

Reply:
I tested the test based on PR #212, it failed 1 out of 10. I see there's difference between upstream test and this one: the former creates the client pod before checking connectivity but the policy is already applied to server pod, so it gives the whole time of creating a pod to enforce policy.
I agree with @antoninbas that we should add a polling with a short timeout (in which we expect the policy must be enforced).
Or we could change the test logic like this:
1. Create a networkpolicy which allows no source (deny all);
2. Create pod A and Pod B and verify they are denied;
3. Update the policy with IPBlock allows the PodCIDR except PodB;
4. Verify A is allowed and B is denied.

Because in step 4, A would do transport layer retry so it waits for policy being enforced implicitly. This would be more like upstream test and doesn't require explicit policy enforcement.

Reply:
Thank you Antonin and Quan for your kind advice ! I think I will go with Quan's test logic since network policy's enforcement time may vary and the length of the timeout may be hard to decide.